### ISSUE TITLE
[Feature request, discussion] Line merging / concatenation

### ISSUE THREAD

**Issue by CarterLi** [fastfetch-cli/fastfetch#482]
## Introduction

As requested in #453, we have several modules prints only a few characters (for example, `Processes` module prints only an integer), it will be nice to merge them into one line.

With the introduction of JSON config, the configuration part for this is possible. I plan to implement Line merging / concatenation in 2.0.

## Configuration

```jsonc
{
  "modules": [
    "os", // Normal module
    {
      "type": "host" // Normal module too
    },

    {
      "type": "~", // A special module type indicates that its child modules should be merged.
      "separator": " | ", // The separator between modules
      "modules": [
        "shell",
        "terminal" // Print Shell and Terminal information in one line, with default key and formation
      ]
    },
    {
      "type": "~",
      "separator": "", // Don't print separator; default value
      "modules": [
        "cpuusage",
        {
          "type": "processes",
          "key": null, // Don't print key
          "format": " (processes: {1})"
        }
      ]
    }
  ]
}
```

Expected result:

```
OS: Arch Linux x86_64
Host: Windows Subsystem for Linux (1.3.11.0)
Shell: fish 3.6.1 | Terminal: Windows Terminal
CPU Usage: 8% (processes: 302)
```

## Design

Principle: line merging should not invade module implementations; that's to say, a module implementation should not care if it's merged or not.

1. Line endings should be handled outside of modules. We print `\n` everywhere, this requires major code refactoring.
2. Logo (cursor moving) should be handled outside of modules. No `ffPrintLogoAndKey`

There are modules prints multiple lines (Display, GPU, Disk, LocalIP, etc). This is what I want to discuss. My basic ideas:

1. Add option `--compact` or `--compact-type` to force them to print in one line, like `--display-compact-type` and `--localip-compact`.
2. In general these modules still print multi lines. However since we are moving `logo` and `line ending` handling outside of modules, it will be problematic when printing second line. Therefore a module should not print anything, but generates a list of strings, so that they don't need to take care of other works between printing two lines.

```c
void ffGenerateGPUOutput(FFGPUOptions* options, FFlist output)
{
  FFlist gpus;
  FF_LIST_FOR_EACH(FFGPUResult, gpu, gpus)
  {
    appendGPUResult(options, index, gpu, (FFstrbuf*) FFListAdd(output));
  }
}
```

3. If we get multiple lines in `~` (line merging) module, just concat them normally.

```jsonc
{
  "modules": [
    {
      "type": "~",
      "separator": " | "
      "modules": [
        "gpu"
      ]
    }
  ]
}
```

Expected result:

```
GPU 1: Intel(R) Iris(R) Xe Graphics (1.00 GiB) [Integrated] | GPU 2: NVIDIA GeForce RTX 3070 Ti Laptop GPU (8.00 GiB) [Discrete]
```

The line may be trancated with `--disable-linewrap` disabled.

## Thoughts?

**Comment by CarterLi** [comment]
Doesn't seem to get any interests. Postponed.

**Comment by tdtooke** [comment]
I probably should have spoken earlier.  I would use this, a little at least.  It makes sense to have shell and terminal on the same line.  I'd probably use it for GPUs as well, I have 2 of them.  A lot of theming stuff occupies multiple lines as well and quite often says the exact same thing.

**Comment by lightproof** [comment]
I'm surprised there is so little feedback on this! I would gladly use it to merge not only shell/terminal, but also DE/WM. And maybe something else.

**Comment by Samueru-sama** [comment]
I would also love this, would use it to merge the Packages and Kernel lines.

**Comment by Paulie420** [comment]
I would also use this... I might be missing other methods of 'renaming' titles of sources or merging sources, but this makes sense to me - or however the developer would like to include these types of features. :P Thanks!!

**Comment by wendellavila** [comment]
I would like to use this as well. Would be great to have cpu + cpuusage on the same line, or icons + cursor, os+kernel...

**Comment by Princess-Sunset-Shimmer** [comment]
i so like this festure,
and i
```jsonc
    {
        "type": "~",
        "separator": " ",
        "modules": [
            {
                "type": "os",
                "key": null,
                "format": "{4}"
            },
            {
                "type": "kernel"
                "key": null
            }
        ]
    }
```
ha ha, then it print
```
```
Blank, literally blank
😫

**Comment by Princess-Sunset-Shimmer** [comment]
> Currently an empty key results in printing the default key, and a `FFstrbuf` can't store a `NULL` value. So users are unable to hide keys. This is problematic.

Soooo, When will it be solved? Tomorrow? TODAY? i got tons of new idea to config fastfetch 20%cooler if you finish adding merging feature to fastfetch

**Comment by timoteuszelle** [comment]
I didn't see it mentioned yet, it would be nice if next to the formatting options already available, to further extend it with regex.
This would allow for more clean and boxed in strings for the exact (wanted) values to be displayed.
Looking at the current use of the formatting options, this might cause some sort of overhaul to accommodate this?

I realize you could load values in scripts and parse those to the jsonc to get the same results.
Thanks for your attention and consideration.

**Comment by spenserblack** [comment]
Just started using `fastfetch` a few minutes ago and already love it!

Might I suggest `["module1", "module2"]` as shorthand `{ "type": "~", "modules": [] }`, with a reasonable default for `separator`?

So it would look like this:
```jsonc
{
  "modules": [
    "os",
    ["shell", "terminal"]
  ]
}
```

and that would be the same as

```jsonc
{
  "modules": [
    "os",
    {
      "type": "~",
      "separator": " | " // or some other default separator
      "modules": ["shell", "terminal"]
    }
  ]
}
```

With a quick glance at the current schema, I don't think this conflicts with anything.

**Comment by tdtooke** [comment]
> Worth mentioning that it can be done right now, but it requires a lot of hacks lol
> 
> This is how I have my config which merges several modules into one line: https://pastebin.com/ntDAPMFT
> 
> I also cache the info of the modules that take a while to load, so that is why some of the modules cat a file in cache.

I tried your method and it indeed works, had to tweak it a bit to make it work for me, I combined Terminal and Terminal Font.  What I did is absolutely horrendous.  I'll share it but don't judge.  I'm not a sed ninja and the pipe heavy way I did this is probably all wrong:
<code>
{
    "key": "│ └",
    "type": "command",
    "keyColor": "blue",
    "text": "( fastfetch -s terminal --pipe -l none; fastfetch -s terminalfont --pipe -l none ) | sed 's/WM //g; s/Theme/-/g' | tr '\n' ' ' | sed 's/Terminal//g' | sed 's/Font//g' | sed 's/󰑃//g' | cut -c 4- | tr -s ' '| sed 's/kitty 0.37.0/&,/g;s/,$//'"
    },</code>

**Comment by Samueru-sama** [comment]
> I tried your method and it indeed works, had to tweak it a bit to make it work for me, I combined Terminal and Terminal Font. What I did is absolutely horrendous. I'll share it but don't judge. I'm not a sed ninja and the pipe heavy way I did this is probably all wron

Yeah I should not have posted a pastebin link, fastfetch has changed and I have had to fix those changes and the output when using the same config broke. 

I have my config up to date here: https://github.com/Samueru-sama/dotfiles/blob/main/Local/config/fastfetch/config.jsonc

I'm no longer merging lines as much as I used to do, and also I have scripts running the background updating every few 15 mins or so that get read by fastfetch instead of typing everything in the config (all of that is in my dotfiles repos).

**Comment by CarterLi** [comment]
Lua or qjs inline script can be used as a work around. Closing.
